src/coreclr/vm/jitinterface.cpp

```diff
@@ -8803,44 +8803,20 @@ CORINFO_CLASS_HANDLE CEEInfo::getDefaultComparerClassHelper(CORINFO_CLASS_HANDLE
     if (Nullable::IsNullableType(elemTypeHnd))
     {
         Instantiation nullableInst = elemTypeHnd.AsMethodTable()->GetInstantiation();
-        TypeHandle iequatable = TypeHandle(CoreLibBinder::GetClass(CLASS__IEQUATABLEGENERIC)).Instantiate(nullableInst);
-        if (nullableInst[0].CanCastTo(iequatable))
-        {
-            TypeHandle resultTh = ((TypeHandle)CoreLibBinder::GetClass(CLASS__NULLABLE_COMPARER)).Instantiate(nullableInst);
-            return CORINFO_CLASS_HANDLE(resultTh.GetMethodTable());
-        }
+        TypeHandle resultTh = ((TypeHandle)CoreLibBinder::GetClass(CLASS__NULLABLE_COMPARER)).Instantiate(nullableInst);
+        return CORINFO_CLASS_HANDLE(resultTh.GetMethodTable());
     }
 
     // We need to special case the Enum comparers based on their underlying type to avoid boxing
     if (elemTypeHnd.IsEnum())
     {
-        MethodTable* targetClass = NULL;
-        CorElementType normType = elemTypeHnd.GetVerifierCorElementType();
-
-        switch(normType)
-        {
-            case ELEMENT_TYPE_I1:
-            case ELEMENT_TYPE_I2:
-            case ELEMENT_TYPE_U1:
-            case ELEMENT_TYPE_U2:
-            case ELEMENT_TYPE_I4:
-            case ELEMENT_TYPE_U4:
-            case ELEMENT_TYPE_I8:
-            case ELEMENT_TYPE_U8:
-            {
-                targetClass = CoreLibBinder::GetClass(CLASS__ENUM_COMPARER);
-                break;
-            }
-
-            default:
-                break;
-        }
+        TypeHandle resultTh = ((TypeHandle)CoreLibBinder::GetClass(CLASS__ENUM_COMPARER)).Instantiate(inst);
+        return CORINFO_CLASS_HANDLE(resultTh.GetMethodTable());
+    }
 
-        if (targetClass != NULL)
-        {
-            TypeHandle resultTh = ((TypeHandle)targetClass->GetCanonicalMethodTable()).Instantiate(inst);
-            return CORINFO_CLASS_HANDLE(resultTh.GetMethodTable());
-        }
+    if (elemTypeHnd.IsCanonicalSubtype())
```

Comment:
The `CanCastTo(IEquatable)` case needs to be after this check, like I have suggested in my comment. 

Or the `IsCanonicalSubtype` check needs to be the very first thing in the method - we would give up on more cases that way.

Or the `CanCastTo(IEquatable)` check needs to be implemented against generic type definition for the `IsCanonicalSubtype` case. This would be the most precise option, but there would be still some ambiguous cases where we would have to give up.

Here is example of a bug that would be hit by the code as written:
```
using System;
using System.Threading;
using System.Collections;
using System.Collections.Generic;
using System.Runtime.CompilerServices;

for (int i = 0; i < 100; i++)
{
   Internal.Console.WriteLine(Test<string, object>().ToString());
   Thread.Sleep(1);
}

[MethodImpl(MethodImplOptions.NoInlining)]
static bool Test<T,U>()
{
    return EqualityComparer<G<T,U>>.Default.Equals(default, default);
}

struct G<T,U> : IEquatable<G<U,T>>
{
   public bool Equals(G<U,T> x) => false;
}
```
It will start printing true correctly, and then flips to printing false as the incorrect Tier1 optimization kicks in.

Reply:
> (It would mirror what happens today.)

It wouldn't, it'd then block `Struct<_Canon>` too, which you were concerned about before.
The issue right now is that this seems to still be devirtualizing `__Canon` as ObjectEqualityComparer for some reason (see [those diffs](https://github.com/dotnet/runtime/pull/88006#discussion_r1251466423)). I have no idea what's going on here, could WellKnownArg be returning some weird handle here?

Reply:
Right, it won't mirror the bugs. Notice that the bug demonstrated by my ` G<T,U>` repro has been shipping for a while (just tried it on .NET 7).

`Struct<_Canon>` needs the third option (use generic type definition to check for IEquatable implementation). Otherwise, it is better to block it rather than to have functional bugs that manifest as observable behavior differences between Tier0 and Tier1 JIT.

Reply:
> see [those diffs](https://github.com/dotnet/runtime/pull/88006#discussion_r1251466423)

This links to my comment. Did you meant to link to some diffs somewhere?

Reply:
> This links to my comment. Did you meant to link to some diffs somewhere?

Yeah, seems like `Copy link` on that GitHub message didn't work so I still had a link to your comment there instead.

Reply:
There diffs look odd. Are you able to reproduce the behavior locally? It can be that the bot picked up stale bits from somewhere.

Reply:
The new diffs look good to me, but I do not think the difference can be explained by the one commit that you have pushed.

Reply:
> The new diffs look good to me, but I do not think the difference can be explained by the one commit that you have pushed.

The bot seems to be having some issues with picking up VM changes, after a fix CoreLib changes look good but frameworks ones still have the weird diff, not sure if it's still an issue with the bot.